derive/tests/client.rs

```diff
@@ -3,43 +3,91 @@ use jsonrpc_core::{IoHandler, Result};
 use jsonrpc_core_client::transports::local;
 use jsonrpc_derive::rpc;
 
-#[rpc]
-pub trait Rpc {
-	#[rpc(name = "add")]
-	fn add(&self, a: u64, b: u64) -> Result<u64>;
+mod client_server {
+	use super::*;
 
-	#[rpc(name = "notify")]
-	fn notify(&self, foo: u64);
-}
+	#[rpc]
+	pub trait Rpc {
+		#[rpc(name = "add")]
+		fn add(&self, a: u64, b: u64) -> Result<u64>;
+
+		#[rpc(name = "notify")]
+		fn notify(&self, foo: u64);
+	}
 
-struct RpcServer;
+	struct RpcServer;
 
-impl Rpc for RpcServer {
-	fn add(&self, a: u64, b: u64) -> Result<u64> {
-		Ok(a + b)
+	impl Rpc for RpcServer {
+		fn add(&self, a: u64, b: u64) -> Result<u64> {
+			Ok(a + b)
+		}
+
+		fn notify(&self, foo: u64) {
+			println!("received {}", foo);
+		}
 	}
 
-	fn notify(&self, foo: u64) {
-		println!("received {}", foo);
+	#[test]
+	fn client_server_roundtrip() {
+		let mut handler = IoHandler::new();
+		handler.extend_with(RpcServer.to_delegate());
+		let (client, rpc_client) = local::connect::<gen_client::Client, _, _>(handler);
+		let fut = client
+			.clone()
+			.add(3, 4)
+			.and_then(move |res| client.notify(res).map(move |_| res))
+			.join(rpc_client)
+			.map(|(res, ())| {
+				assert_eq!(res, 7);
+			})
+			.map_err(|err| {
+				eprintln!("{:?}", err);
+				assert!(false);
+			});
+		tokio::run(fut);
 	}
 }
 
-#[test]
-fn client_server_roundtrip() {
-	let mut handler = IoHandler::new();
-	handler.extend_with(RpcServer.to_delegate());
-	let (client, rpc_client) = local::connect::<gen_client::Client, _, _>(handler);
-	let fut = client
-		.clone()
-		.add(3, 4)
-		.and_then(move |res| client.notify(res).map(move |_| res))
-		.join(rpc_client)
-		.map(|(res, ())| {
-			assert_eq!(res, 7);
-		})
-		.map_err(|err| {
-			eprintln!("{:?}", err);
-			assert!(false);
+mod named_params {
+	use super::*;
+	use jsonrpc_core::Params;
+	use serde_json::json;
+
+	#[rpc(client)]
+	pub trait Rpc {
+		#[rpc(name = "call_with_named", named_params)]
```

Comment:
Would it make sense to specify `named_params` in the trait level attribute instead? On the one hand it's good to have the fine grained control - but in practice wouldn't all methods be either named or positional?

Reply:
We already have `raw_params` on the method level. I think this is a good idea to have it here, but an additional attribute on the trait level would be cool tool, especially in the future if we implement server side as well.